be/src/exprs/math_functions.cpp

```diff
@@ -899,10 +898,310 @@ StatusOr<ColumnPtr> MathFunctions::cosine_similarity(FunctionContext* context, c
     return result;
 }
 
-// explicitly instaniate template function.
+template <LogicalType TYPE, bool isNorm>
+StatusOr<ColumnPtr> MathFunctions::cosine_similarity2(FunctionContext* context, const Columns& columns) {
+    DCHECK_EQ(columns.size(), 2);
+
+    const Column* base = columns[0].get();
+    const Column* target = columns[1].get();
+
+    if (base->has_null() || target->has_null()) {
+        return Status::InvalidArgument(
+                fmt::format("cosine_similarity does not support null values. {} array has null value.",
+                            base->has_null() ? "base" : "target"));
+    }
+    if (base->is_nullable()) {
+        base = down_cast<const NullableColumn*>(base)->data_column().get();
+    }
+    if (target->is_nullable()) {
+        target = down_cast<const NullableColumn*>(target)->data_column().get();
+    }
+
+    if (!base->is_constant() && !target->is_constant()) {
+        return Status::InvalidArgument(
+                fmt::format("cosine_similarity need a constant array as one of the parameter!"));
+    }
+
+    using CppType = RunTimeCppType<TYPE>;
+    using ColumnType = RunTimeColumnType<TYPE>;
+
+    std::vector<CppType> base_elements;
+
+    auto pattern = ColumnHelper::get_const_value<TYPE_VARCHAR>(base);
+
+    std::string pattern_str = pattern.to_string();
+    std::string delimiter = ",";
+    size_t pos_start = 0, pos_end, delim_len = delimiter.length();
+    std::string token;
+    while ((pos_end = pattern_str.find(delimiter, pos_start)) != std::string::npos) {
+        token = pattern_str.substr(pos_start, pos_end - pos_start);
+        CppType num = ::atof(token.c_str());
```

Comment:
What if `pattern_str` is not with valid format? such as `abc, 123, 456`, `1.2.3, 456`, `1e10000`.

Reply:
If `pattern_str` is not in a valid format, such as "abc, 123, 456, 1.2.3, 456, 1e10000", then the ::atof(token.c_str()) function will attempt to convert the string into a floating-point number. If the string cannot be parsed into a valid floating-point number, the ::atof() function will return 0.0. For "abc" and "1.2.3", ::atof() will return 0.0. For "1e10000", since this number exceeds the range of a floating-point number, ::atof() will also return 0.0.